### PR TITLE
Patch to optionally ignore the XID field in OFMessage and other types

### PR DESCRIPTION
**Pull request by rizard**
Reviewer: @rlane @andi-bigswitch 

For any class with an XID field (detected by a class field named 'xid'), include an equalsIgnoreXid() function to ignore the XID in the equality check. The assumption, of course, is that the XID field is named 'xid' and is a primitive type ('long' at present). This will allow for the content-comparison of two OFMessages without regard to when they were created (or what the XID was set to).

This will address the long-lived problem in Floodlight of not being able to use the OFMessageDamper to prevent rapid, duplicate messages from being sent to the switch. A workaround has been used the last couple years that rebuilds the 'other' object in the comparison with the XID of 'this', which is quite inefficient.

### CODE REVIEW

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/pr-493-jenkins-loxigen_pull_req_artifacts-27

**Comment by bsn-abat** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.bigswitch.com/job/loxigen_pull_req_artifacts/27/

**Comment by bsn-abat** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.bigswitch.com/job/loxigen_pull_req_docker/32/

File: java_gen/templates/of_class.java [hunk 1]
```diff
@@ -419,6 +419,42 @@ public boolean equals(Object obj) {
         return true;
     }
 
+    //:: for prop in msg.data_members:
```
Comment (andi-bigswitch) [inline java_gen/templates/of_class.java]: Weird loop here - could use 
```python
  if find(lambda m: m.name == "xid", msg.data_members):
```

File: java_gen/templates/of_class.java [hunk 2]
```diff
@@ -419,6 +419,42 @@ public boolean equals(Object obj) {
...
+        } else if (!${prop.name}.equals(other.${prop.name}))
+            return false;
+        //:: #endif
+        //:: #endfor
+        return true;
+    }
+    //:: break
+    //:: #endif
+    //:: #endfor
+
```
Comment (andi-bigswitch) [inline java_gen/templates/of_class.java]: Nit - remove this newline to avoid changing objects that don't have a XID (see artifact link).

**Comment by andi-bigswitch** [comment]
LGTM otherwise

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/pr-493-jenkins-loxigen_pull_req_artifacts-31

**Comment by bsn-abat** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.bigswitch.com/job/loxigen_pull_req_artifacts/31/

**Comment by bsn-abat** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.bigswitch.com/job/loxigen_pull_req_docker/36/

File: java_gen/templates/of_class.java (hunk 2 above)
Comment (rizard) [inline java_gen/templates/of_class.java]: @andi-bigswitch not sure I understand this one. Can you elaborate? I see only a single newline separating the functions, which appears to exist b/t all functions defined in of_class.java.

File: java_gen/templates/of_class.java (hunk 1 above)
Comment (rizard) [inline java_gen/templates/of_class.java]: @andi-bigswitch agreed and fixed

File: java_gen/templates/of_class.java (hunk 2 above)
Comment (andi-bigswitch) [inline java_gen/templates/of_class.java]: Sure; it's really a small nit:

When you look at the artifact delta created from this pull request (the latest commit in the artifact repo that bsn-abat links to, https://github.com/floodlight/loxigen-artifacts/commit/2e58a1ba4849e79e104158d4db47b4651f0666d0), you'll see that *all* classes gain an additional newline (not just the ones that actually have an xid field). The reason is that the newline in L.457 (it has moved up a couple of lines with your latest commit) gets inserted unconditionally.

TL;DR: if you move the newline into the #if/#endif block it should be correct.

Comment (rizard) [inline java_gen/templates/of_class.java]: @andi-bigswitch Oh, wow...that's very obvious now. Thanks for the clarification! I should have caught that.

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/pr-493-jenkins-loxigen_pull_req_artifacts-33

**Comment by bsn-abat** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.bigswitch.com/job/loxigen_pull_req_artifacts/33/

**Comment by bsn-abat** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.bigswitch.com/job/loxigen_pull_req_docker/38/

**Comment by rizard** [comment]
@andi-bigswitch I've made the requested changes. Let me know if there's anything else you'd like tweaked

**Comment by andi-bigswitch** [comment]
LGTM

ACK

**Comment by bsn-abat** [comment]
ABAT: START_MERGE:  Log file at http://vss1/abat/2016.06.09.1351-m.loxigen.master/abat.log

**Comment by bsn-abat** [comment]
ABAT: ACCEPT:  Successfully merged

In case you want to see the build log, check out :
Log file at http://vss1/abat/2016.06.09.1351-m.loxigen.master/abat.log
/cc
